e3nn/tensor_product.py

```diff
@@ -73,39 +73,70 @@ def forward(self, features_1, features_2):
 
 
 class LearnableTensorProduct(torch.nn.Module):
-    def __init__(self, Rs_mid_1, Rs_mid_2, mul_mid, Rs_out, selection_rule=o3.selection_rule):
+    def __init__(self, Rs_mid_1, Rs_mid_2, Rs_out,
+                 selection_rule=o3.selection_rule, groups=1):
         super().__init__()
-        self.mul_mid = mul_mid
+        self.groups = groups
         self.tp = TensorProduct(Rs_mid_1, Rs_mid_2, selection_rule)
-        self.lin = Linear(mul_mid * self.tp.Rs_out, Rs_out)
+        self.lin = Linear(groups * self.tp.Rs_out, Rs_out)
 
     def forward(self, x1, x2):
         """
         :return:         tensor [..., channel]
         """
         # split into mul x Rs
         *size, _ = x1.shape
-        x1 = x1.view(*size, self.mul_mid, -1)
-        x2 = x2.view(*size, self.mul_mid, -1)
+        x1 = x1.view(*size, self.groups, -1)
+        x2 = x2.view(*size, self.groups, -1)
 
         x = self.tp(x1, x2).view(*size, -1)
         x = self.lin(x)
         return x
 
 
 class LearnableBispectrum(torch.nn.Module):
-    def __init__(self, Rs_in, mul_hidden, mul_out):
+    def __init__(self, Rs_in, Rs_hidden, mul_out, lmax=None):
         super().__init__()
-        self.lmax = max(l for mul, l in Rs_in)
-        Rs_hidden = [(mul_hidden, l) for l in range(self.lmax + 1)]
+        self.Rs_in = rs.simplify(Rs_in)
+        self.Rs_hidden = rs.simplify(Rs_hidden)
+        if lmax is None:
+            self.lmax = max(l for mul, l, _ in self.Rs_hidden)
+        else:
+            self.lmax = lmax
         # Learnable tensor product of signal with itself
-        self.tp = LearnableTensorProduct(Rs_in, Rs_in, 1, Rs_hidden,
-                                         partial(o3.selection_rule, lmax=self.lmax))
+        self.tp = LearnableTensorProduct(
+            self.Rs_in, self.Rs_in, self.Rs_hidden,
+            partial(o3.selection_rule, lmax=self.lmax), groups=1)
```

Comment:
The selection of the `l` that are present in the output of the `LearnableTensorProduct` should be done automatically inside `LearnableTensorProduct`. Because never you will allow a l that is not in the output rep. It will actually raise an error in `Linear`.